### ISSUE TITLE
Potential inconsistencies between the official implementation and the paper

### ISSUE THREAD

**Issue by ranqianQQ** [zzn-promax/CMGNN#3]
Hi, thank you for releasing the official implementation of CMGNN.

While reviewing and reproducing the code, I noticed several potential inconsistencies between the released implementation and the methodology described in the paper. I would appreciate it if the authors could help clarify whether these behaviors are expected and whether the released code is the exact version used for the reported experiments.

## 1. `M_fix` appears to use the ground-truth labels of all nodes

In the current implementation, `M_fix` appears to be calculated using the complete ground-truth label vector, including the labels of validation and test nodes.

Moreover, `M_fix` is directly used for message passing during training.

However, according to Eqs. (10)–(12), under the semi-supervised setting, the compatibility matrix should be estimated using:

- ground-truth labels for training nodes;
- model predictions or pseudo-labels for non-training nodes;
- confidence weights derived from prediction entropy.

Therefore, using validation and test ground-truth labels to construct `M_fix` may introduce label leakage, because information from the evaluation sets directly participates in training-time message passing.

Could you please confirm:

1. Whether `M_fix` was used in the experiments reported in the paper.
2. Whether validation and test labels are intentionally included when calculating `M_fix`.
3. Whether non-training nodes should instead use model predictions or pseudo-labels.

## 2. The discrimination loss is calculated but not included in the optimization objective

The implementation calculates the discrimination loss, but it appears that only the cross-entropy loss is used for backpropagation.

The current behavior seems equivalent to:

```python
loss_dis = discrimination_loss(...)
loss = loss_ce
loss.backward()
```

However, Eq. (17) defines the complete optimization objective as:

```text
L = L_ce + lambda * L_dis
```

Therefore, if `loss_dis` is calculated but not added to the final loss, it will not contribute any gradients and will not affect model training.

The expected behavior should be similar to:

```python
loss = loss_ce + lambda_dis * loss_dis
loss.backward()
```

This also seems inconsistent with the ablation study involving the discrimination loss (`W/O DL`).

Could you please confirm whether the discrimination loss was accidentally omitted from the released implementation, or whether another version of the code was used for the reported experiments?

## 3. Local prototype indices appear to be used as global node indices

When constructing class prototypes, the implementation appears to obtain indices after selecting the training subset.

For example:

```python
local_indices = torch.where(labels[train_mask] == class_id)[0]
```

The resulting indices are local positions within `labels[train_mask]`, rather than global node indices in the original graph.

If these local indices are subsequently used to index the original feature matrix:

```python
prototype = features[local_indices].mean(dim=0)
```

the selected nodes may not correspond to the actual training nodes belonging to the target class.

A possible correct implementation would be:

```python
train_indices = torch.where(train_mask)[0]
class_indices = train_indices[labels[train_indices] == class_id]
prototype = features[class_indices].mean(dim=0)
```

Alternatively:

```python
prototype = features[train_mask][
    labels[train_mask] == class_id
].mean(dim=0)
```

According to Eq. (9), class prototypes should be constructed from the features and labels of the actual training nodes.

Could you please check whether local subset indices are being incorrectly treated as global graph indices in the current implementation?

## 4. The degree-weight conditions overlap and do not exactly follow Eq. (27)

Eq. (27) defines the degree weight using the following piecewise function:

```text
w_i^d = d_i / (2K),          if d_i <= K
        0.25 + d_i / (4K),   if K < d_i <= 3K
        1,                    otherwise
```

The first and second intervals should be mutually exclusive.

However, in the current implementation, the two conditions appear to overlap. For example, if the conditions are implemented as:

```python
low_mask = degree <= K
middle_mask = degree <= 3 * K
```

then every node satisfying `degree <= K` also satisfies `degree <= 3K`.

If the second assignment is executed after the first assignment, the value calculated by the first branch will be overwritten.

A non-overlapping implementation should be similar to:

```python
degree_weight = torch.ones_like(degree, dtype=torch.float)

low_mask = degree <= K
middle_mask = (degree > K) & (degree <= 3 * K)

degree_weight[low_mask] = degree[low_mask] / (2 * K)
degree_weight[middle_mask] = (
    0.25 + degree[middle_mask] / (4 * K)
)
```

Could you please confirm whether the overlapping conditions in the released implementation are an implementation error?

## Summary

These potential issues may affect the reproducibility and correctness of the reported results:

1. Validation and test ground-truth labels may be used during training through `M_fix`.
2. The discrimination loss may be calculated but not used for optimization.
3. Class prototypes may be constructed using incorrect node indices.
4. The degree-weight implementation may not match the non-overlapping intervals in Eq. (27).

Could you please clarify whether the released code is the exact implementation used to obtain the results reported in the paper?

If these are implementation issues, it would be helpful if the authors could provide a corrected version or identify the commit and configuration used for the reported experiments.

Thank you for your time.

**Comment by zzn-promax** [comment]
Thank you very much for your careful review and for pointing out these issues.

After checking the released files, we found that a version used for one of our ablation studies was mistakenly uploaded instead of the complete implementation used for the main experiments. We sincerely apologize for the confusion this has caused.

We are currently auditing the implementation, correcting the identified issues, and rerunning the experiments to verify the results. Once this process is complete, we will update the repository with the corrected code and the corresponding experimental configurations.

We will also provide a more detailed point-by-point response after completing the verification. Thank you again for helping us identify this problem and improve the reproducibility of our work.

**Comment by zzn-promax** [comment]
Thank you again for your careful review and for your patience. We have now completed the code audit and updated the implementation and dataset-specific configurations in commit [580dcb4](https://github.com/zzn-promax/CMGNN/commit/580dcb42542ba52119187501beeb7b5879cd05b0).

Regarding the four issues you identified:

1. **Compatibility matrix and label leakage**

   You are correct that `M_fix` in the previously released version was calculated using the complete label vector. This code came from an ablation-study version that was mistakenly uploaded. We have removed `M_fix`. In the corrected implementation, ground-truth labels are used only for the training nodes, while predictions are used for the remaining nodes. Entropy-based confidence and degree weights are then applied when estimating the compatibility matrix. No validation or test labels are used in this process.

2. **Discrimination loss**

    The previously uploaded ablation-study version excluded the discrimination loss from the final objective. It is now correctly included together with the cross-entropy loss, consistent with the complete model described in the paper.

3. **Prototype node indexing**

   The local/global indexing error has been corrected. Class prototypes are now computed from the actual training nodes belonging to each class. When a class has no labeled training samples in a split, its corresponding virtual node is excluded from the supervised training mask. We also provide a configurable smoothing coefficient `r` for prototype construction, with the dataset-specific values included in `config/CMGNN.yaml`.

4. **Degree-based weighting**

   The degree masks have been changed to mutually exclusive intervals, consistent with Eq. (27) in the paper.

We have also rerun our implementation checks and verified that the updated code can complete both training and inference under the feature-only and structure-information settings.

We sincerely apologize for the confusion caused by the previously uploaded version. Thank you again for identifying these problems and helping us improve the correctness and reproducibility of the repository. Please let us know if you notice any further inconsistencies.
